### DISCUSSION TITLE
Is is possible to attach methods to models in Prisma?

### DISCUSSION THREAD

**Discussion by florianmartens** [prisma/prisma#2763] [Q&A] [answered]
Hi,

so, a pretty common pattern for ORMs is that you can typically attach methods to every instance of a model. In type ORM you'd do that via the entity:

```
@Entity()
export class User extends BaseEntity {
    @Column()
    username: string;

    async validatePassword(password: string) {
        ...
    }
}
```

Is this possible for Prisma, too? I would be very handy to be able to have every instance of a model have some sort of method attached to it.

**Comment by ryands17** [comment, marked as answer]
Hey @florianmartens :wave: 
As Prisma is not an ORM as described [here](https://www.prisma.io/docs/understand-prisma/prisma-in-your-stack/is-prisma-an-orm), it just returns a plain JS object on querying. So it's not possible to add methods on the model as there is no Instance, it's just a plain object that's returned.

**Reply by florianmartens** [reply]
@ryands17 thanks for your answer! Is there another pattern to achieve somewhat the same thing (have a method attached to every returned JS object of a model)? Or is this an anti-pattern in Prisma?

**Reply by ryands17** [reply]
One way would be creating a custom class in which you require the methods, query data in those methods via Prisma and add your own methods on the class and use that Instance (something like Active Record). I don't think it would be advisable to monkey-patch custom methods on the Prisma created methods.

**Comment by tobiasdiez** [comment]
I have a similar issue. The following works but is relatively complicated. @ryands17 is this what you had in mind? Is there really no easier solution? It seems like a common requirement if you don't just want to have intelligent domain objects.

```typescript
import { User as UserModel } from '@prisma/client'

interface User extends UserModel {
  verifyPassword (): boolean
}

map (user: UserModel): User {
    return { ...user, verifyPassword: () => { return false } }
}

function(...) {
   const user = await prisma.user.findUnique({...})
   const userExt = map(user)
   userExt.verifyPassword()
}
```

**Reply by ryands17** [reply]
Currently this would be the only way to do so. We have a request for the same [here](https://github.com/prisma/prisma/issues/5998) so we will be looking into this.

**Comment by edu791** [comment]
Is there any solution nowadays with Prisma 2?

**Reply by ryands17** [reply]
@edu791 Currently there's no solution to attach methods. You can follow these requests for more updates:
https://github.com/prisma/prisma/issues/5998
https://github.com/prisma/prisma/issues/5273

**Reply by edu791** [reply]
@ryands17 Is there a plan in the future to support mapping directly to class instances instead of plain objects?

**Reply by ryands17** [reply]
Not in the near future though as per our roadmap: https://www.notion.so/prismaio/Prisma-public-roadmap-50766227b779464ab98899accb98295f

**Comment by dcsan** [comment]
another hacky way around this is a wrapper class that contains the prisma object as an internal `data` field.

but this means none of the prisma methods are exposed on the wrapper class, so a lot of tedious mapping/duplication is needed.

```ts
export class CPlayer {
    data?: Player | null

    constructor(data?: Player | null) {
        this.data = data
    }

    static async findOrCreate(query: Prisma.PlayerWhereInput): Promise<CPlayer> {
        let data = await prisma.player.findFirst({ where: query })
        if (!data) {
            const input = query as Prisma.PlayerCreateInput
            data = await prisma.player.create({ data: input })
        }
        // construct object and return instance
        return new CPlayer(data)
    }

    // move a player
    async moveTo(px: number, py: number) {
        this.data = await prisma.player.update({
            where: { id: this.data!.id },
            data: { px, py },
        })
        console.log('new data', this.data)
    }

}

// use it like
const player = CPlayer.findOrCreate({name: 'bob'})

```

**Comment by olance** [comment]
Came to find the same kind of tricks!

> but this means none of the prisma methods are exposed on the wrapper class

What about using Proxy?

```js
class BaseModel {
    static makeModel(modelData) {
        return new Proxy(modelData, new this);
    }
    
    get(target, property) {
        if (property in this) {
            return this[property];
        }
        
        return target[property];
    }

    has(target, property) {
        return (property in this) || (property in target)
    }
}

export UserModel extends BaseModel {
    verifyPassword(providedPassword) {
        // Don't do this at home!
        return this.password === providedPassword;
    }
}
```

The `BaseModel` class provides a convenience method that initializes a `Proxy` wrapping the Prisma "model" object, with a new instance of itself as the proxy handler.

You'd use it this way:

```js
const user = UserModel.makeModel(await prisma.user.findUnique({...}))
if (!user.verifyPassword(providedPassword)) {
    throw new Error("Invalid password");
}
```

Curious to get your feedback on this; will try to start using this in a project of mine!

**Comment by dcsan** [comment]
using proxy is certainly an interesting idea, but it seems quite an ugly API into instance creation, wrapping and passing an already queried prisma object:

```ts
const user = UserModel.makeModel(await prisma.user.findUnique({ name: 'bob' }))
```

I would prefer if there was a way to add all the prisma methods to the User class so we can do like this

```ts
const user = UserModel.makeModel({ name: 'bob' })

// or
const user = UserModel.findUnique({ name: 'bob' })

```

and all other methods could then be patched onto the User class.


A problem I ran into before is that all the prisma model generated types are unique, there is no hierarchy or baseClass or common interface for them. eg these are all unique and share nothing:
`Prisma.PlayerWhereInput`
`Prisma.UserWhereInput`

So there wasn't any way to have my own BaseClass that would implement the finder methods in a generic way, so every class I create has to have a ton of boilerplate, basically all the methods all the time.

There maybe a way with metaprogramming to generate all those methods, but this is more fragile magic piled on top of an already opaque build prisma generator process, and feels hard to maintain.

**Reply by olance** [reply]
You're right, the API for my proposal isn't great.
But something like `UserModel.makeModel({ name: 'bob' })` is ambiguous: is that a new record creation, or trying to find something and it's a `where` object, etc.

Maybe what's needed is a wrapper of Prisma's client methods at the model level, like:

```js
const user = await UserModel.findUnique({ name: 'bob' })
```

Interestingly that's [what Redwoodjs is exploring](https://redwoodjs.com/docs/redwoodrecord) with their RedwoodRecord layer on top of Prisma!

**Reply by dcsan** [reply]
yeah that makes sense those github guys were old ruby on rails heads, and the API looks really like active record. I also find it much less boilerplate to work with. But an ORM built on top of prisma which is already a ton of magic... better hope nothing breaks cos it will be hell trying to unravel it.

**Reply by olance** [reply]
Indeed it's heavily inspired by RoR – as a former rubyist I like it a lot 😅 

RoR is already a ton of magic, as you gain experience working with it you start understanding the magic behind it.
I guess it can be true for any system.

And in any case, your remark stands for most famous JS framework/libs nowadays, so I'd rather choose the best DevX to work with than the least magical, but that's just my personal take of course 😊

**Comment by fred-10git** [comment]
First of all, i'm a really newbie in JS (i've just discovered proxies^^), TS and prisma ^^

But found this post (and others)  trying to mix differents methods to make my own trick (seems working as i want, i have to try furthermore), using an intern data object (prisma result) and two layers of proxies : [prismaModel] > [data] > [class handler]

```
import { PrismaClient } from "@prisma/client";

const prismaClient = new PrismaClient();

const proxyCreate = (handler: any) => ({
  get: (target: any, key: string) => {
    if (key in target) {
      return target[key as keyof typeof target];
    } else if (key in handler) {
      return handler[key as keyof typeof handler];
    } else {
      return undefined;
    }
  },

  has: (target: any, key: string) => {
    return key in target || key in handler;
  },

  set: (target: any, key: any, value: any) => {
    if ("data" in handler) {
      handler["data"][key] = value;
    } else {
      target[key as keyof typeof target] = value;
    }
    return true;
  },
});

class Model {
  data: any;
  proxy: any;

  // - Have to add those because of TS :
  findFirst: Function = () => {};
  findMany: Function = () => {};
  findUnique: Function = () => {};
  create: Function = () => {};
  // ... to be continued

  constructor(prismaClient: any, obj: any = {}) {
    this.data = obj;
    this.proxy = new Proxy(prismaClient, proxyCreate(this.data));
  }

  async save() {
    return await this.create({ data: this.data });
  }
}

export default class User extends Model {
  // - Have to add those beacuse of TS for using them direct on User object:
  id: unknown = null;
  username: unknown = "";
  email: unknown = "";
  password: unknown = "";

  constructor(obj: any = {}) {
    super(prismaClient.userModel, obj);
    return new Proxy(this.proxy, proxyCreate(this));
  }

  static async myFindFirst() {
    let user = new User();
    user = new User(await user.findFirst());
    return user;
  }
}

```

Then, in use :

```
import { User } from "./models";

const main = async () => {
  // - Getting a User
  const user: User = await User.myFindFirst();
  console.log("Direct access to data values : ", user.id, user.username);

  // - Using directly User object to access prisma Model functions
  const doc = await user.findFirst({ where: { email: "toto@gmail.com" } });
  console.log(doc);

  //- Create user :
  const newUser = new User();
  newUser.username = "bob";
  newUser.email = "bob@gmail.com";
  newUser.password = "123";
  await newUser.save();
};

main();
```

What do you think ? 
As i'm totally new in TS, I have to add fake properties and methods (and giving priority to target in my proxy function), to avoid adding @ts-ignore each time i access something on user that is on underlayer proxies : maybe there's a better way ?

**Comment by angelhdzmultimedia** [comment]
Any update?

### Update

Found this https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions
Thanks.

**Comment by anatolzak** [comment]
**UPDATE**

Adding a `validatePassword` method to a user can be easily accomplished by [adding custom methods to a query's result](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/result).

```typescript
import { PrismaClient } from '@prisma/client';
import bcrypt from 'bcryptjs';

const prisma = new PrismaClient().$extends({
  result: {
    user: {
      validatePassword: {
        needs: { password: true },
        compute: user => async (rawPassword: string) => {
          return await bcrypt.compare(rawPassword, user.password);
        },
      },
    },
  },
});

async function main() {
  const user = await prisma.user.findFirst({ where: { email: 'test@gmail.com' } });

  if (user) {
    const isValidPassword = user.validatePassword('mypassword');
  }
}
```

And you can also automatically hash the password when creating or updating a user using [custom Prisma client queries](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query)

```typescript
import { PrismaClient, Prisma } from '@prisma/client';

const prisma = new PrismaClient().$extends({
  query: {
    user: {
      async $allOperations({ args, model, operation, query }) {
        if ((operation === 'create' || operation === 'update') && 'data' in args) {
          const data = args.data as Prisma.UserCreateInput | Prisma.UserUpdateInput;

          if (data.password) {
            data.password = hashPassword(data.password);
          }
        }

        return query(args);
      },
    },
  },
});
```

Credit to @angelhdzmultimedia

**Comment by angelhdzmultimedia** [comment]
> This solution isn't perfect, but it gets you close to what you're aiming to do by utilizing [Prisma's custom models](https://www.prisma.io/docs/concepts/components/prisma-client/custom-models).
> 
> In the script below, a new method called `validatePassword` is added to the `user` model through Prisma's `$extend` function. This method is designed to take an email and a password as its parameters:
> 
> ```ts
> import { PrismaClient } from '@prisma/client'
> 
> const prisma = new PrismaClient().$extend({
>   model: {
>     user: {
>       async validatePassword(email: string, password: string) {
>         // Your validation logic here
>       },
>     },
>   },
> })
> 
> async function main() {
>   await prisma.user.validatePassword('user2@example2.com', 's3cret')
> }
> ```
> 
> It's important to note that `validatePassword` is a static method, meaning it is called on the `user` model itself and not on an instance of a `user`. Prisma operates with plain objects rather than class instances, so adding methods to individual user instances (like you might with other ORMs) isn't an option here. However, by using a static method with Prisma's custom models, you can still organize and centralize logic such as password validation in a coherent manner, providing a structured way to validate user credentials by simply passing the email and password to the `validatePassword` method.

You can add the validatePassword method as a result extension. Prisma extensions can be made for:

![image](https://github.com/prisma/prisma/assets/10191085/f2904587-826c-404d-9c1a-85eafcd1b3f4)

I add my validatePassword as a `result` extension, and then use a `query` extension to override the create query to hash the password before saving it in the db, and excluding the password from both extensions, so any find or create query, will not have the user's password.

So I would define a login.prisma.ts extension and a register.prisma.ts extension with the certainity that using these clients won't include the user's password.

```ts
const user = await loginPrisma.user.findOne({email})

if (!user) {
  throw new NotFoundException('Email not found')
}

const isPasswordValid: boolean = await user.validatePassword( loginDto.password )

if (!isPasswordValid) {
throw new UnauthorizedException('Email or password not valid')
}

return user  // no password!
```

**Comment by ludralph** [comment]
Hi @florianmartens,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
